File: examples/oshmem_symmetric_data.java
import shmem.*;
import java.nio.*;

public class oshmem_symmetric_data
{
    private static final int SIZE = 16;

    public static void main(String[] args) throws ShMemException
    {
        ShMem.startPEs(0);

        int numPE = ShMem.getNumPEs(),
            myPE  = ShMem.getMyPE();

        int[] source = new int[SIZE];
        Addr  target = new Addr(4 * SIZE); // int is 4 bytes

        if(myPE == 0)
        {
            // initialize array
            for(int i = 0; i < SIZE; i++)
                source[i] = i;

            // local, not symmetric
            // static makes it symmetric
            // put "size" words into target on each PE
            for(int i = 1; i < numPE; i++)
                target.putInt(source, i);
        }

        ShMem.barrierAll(); // sync sender and receiver

        if(myPE != 0)
        {
            StringBuilder sb = new StringBuilder();
            sb.append("Target on PE "+ myPE +" is \t");
            IntBuffer buf = target.asIntBuffer();

            for(int i = 0; i < SIZE; i++)
                sb.append(buf.get(i) +" \t");

            sb.append('\n');
            System.out.print(sb);
        }

        ShMem.barrierAll(); // sync before exiting
        target.free();
    }
}

File: examples/oshmem_symmetric_data.c
/*
 * Copyright (c) 2014-2016 Mellanox Technologies, Inc.
 *                         All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include <stdio.h>
#include <shmem.h>

#define SIZE 16

int main(int argc, char* argv[])
{
    short source[SIZE];
    static short target[SIZE];
    int i;
    int num_pe, my_pe;

    shmem_init();

    num_pe = shmem_n_pes();
    my_pe = shmem_my_pe();

    if (my_pe == 0) {
        /* initialize array */
        for(i = 0; i < SIZE; i++) {
            source[i] = i;
        }
        /* local, not symmetric */
        /* static makes it symmetric */
        /* put "size" words into target on each PE */
        for(i = 1; i < num_pe; i++) {
            shmem_short_put(target, source, SIZE, i);
        }
    }

    shmem_barrier_all(); /* sync sender and receiver */

    if (my_pe != 0) {
        printf("Target on PE %d is \t", my_pe);

        for(i = 0; i < SIZE; i++) {
            printf("%hd \t", target[i]);
        }
        printf("\n");
    }

    shmem_barrier_all(); /* sync before exiting */
    shmem_finalize();

    return 0;
}
